File: 2018/day01/day01.py
def read_input() -> str:
    try:
        f = open("./input.txt", "r")
        # f = open("./example.txt", "r")
        return f.read()
    except OSError:
        exit(1)


def part_one():
    result = sum([int(x) for x in read_input().split("\n")])

    print(f"Part One: {result}")


def part_two():
    input = [int(x) for x in read_input().split("\n")]
    result = None
    frequency = 0
    mem = {}

    while result == None:
        for x in input:
            frequency += x
            if frequency in mem:
                result = frequency
                break
            else:
                mem[frequency] = 1

    print(f"Part Two: {result}")


if __name__ == "__main__":
    part_one()
    part_two()

File: 2015/day01/day01.hs
import Data.Array (listArray)

transform :: Char -> Int
transform x
  | x == '(' = 1
  | x == ')' = - 1
  | otherwise = 0

transformString :: String -> [Int]
transformString = map transform

firstNegativePos :: String -> Int
firstNegativePos input = do
  -- This probably can be simplified by dots and dollar signs, but didn't quite got how it works yet
  let a = transformString input
  let b = scanl (+) 0 a
  let d = takeWhile (>= 0) b
  length d

partOne :: String -> IO ()
partOne input = do
  putStrLn $ "Part One: " ++ show (sum $ transformString input)

partTwo :: String -> IO ()
partTwo input = do
  putStrLn $ "Part Two: " ++ show (firstNegativePos input)

main :: IO ()
main = do
  input <- readFile "./input.txt"

  partOne input
  partTwo input